driver: sensor: Add support for LIS2MDL Mag sensor

mariotesi opened the pull request:
Added support to LIS2MDL Magnetometer sensor provided
with following features:
	- I2C interface
	- Mag data
	- Temperature data
	- ODR configurable by config or at runtime
	- Trigger mode selectable by menuconfig
	- IRQ pin configurable
	- Hard Iron offset setting at runtime

Tested on ST MEMS IKS01A2 + NUCLEO STM32F411RE board.
LIS2MDL connected to I2C master interface (SPI 3 wire not
supported yet).
Test run with all ODR {10, 20, 50, 100} Hz in poll and
trigger mode.

For more info on this LIS2MDL please follow this link:
http://www.st.com/en/mems-and-sensors/lis2mdl.html

Signed-off-by: Mario Tesi <mario.tesi@st.com>

mariotesi committed:
driver: sensor: Add support for LIS2MDL Mag sensor

Added support to LIS2MDL Magnetometer sensor provided
with following features:
	- I2C interface
	- Mag data
	- Temperature data
	- ODR configurable by config or at runtime
	- Trigger mode selectable by menuconfig
	- IRQ pin configurable (by dts or Kconfig)
	- Hard Iron offset setting at runtime
	- Include yaml file

Tested on ST MEMS IKS01A2 + NUCLEO STM32F411RE board.
LIS2MDL connected to I2C master interface (SPI 3 wire not
supported yet).
Test run with all ODR {10, 20, 50, 100} Hz in poll and
trigger mode.

GPIO IRQ dts configuration has been tested by adding
to boards/arm/nucleo_f411re/nucleo_f411re.dts file
this patch:

&i2c1 {
        status = "ok";
        clock-frequency = <I2C_BITRATE_FAST>;
+
+       /* ST Microelectronics LIS2MDL mag sensor */
+       lis2mdl-magn@1e {
+               compatible = "st,lis2mdl-magn";
+               reg = <0x1e>;
+               irq-gpios = <&gpioa 4 0>;
+               label = "LIS2MDL";
+               status = "ok";
+       };
};

and adding boards/arm/nucleo_f411re/dts.fixup with following
content:

	ST_STM32_I2C_V1_40005400_ST_LIS2MDL_MAGN_1E_LABEL
	ST_STM32_I2C_V1_40005400_ST_LIS2MDL_MAGN_1E_BASE_ADDRESS
	ST_STM32_I2C_V1_40005400_ST_LIS2MDL_MAGN_1E_BUS_NAME
	ST_STM32_I2C_V1_40005400_ST_LIS2MDL_MAGN_1E_IRQ_GPIOS_CONTROLLER
	ST_STM32_I2C_V1_40005400_ST_LIS2MDL_MAGN_1E_IRQ_GPIOS_PIN

For more info on this LIS2MDL please follow this link:
http://www.st.com/en/mems-and-sensors/lis2mdl.html

Signed-off-by: Mario Tesi <mario.tesi@st.com>

galak reviewed (changes requested):
Need device tree binding/yaml for these sensor.  Use compatiable “st,lis2mdl”.  See dts/bindings/sensor/st,lis3mdl-magn.yaml for example.

erwango commented on drivers/sensor/lis2mdl/Kconfig:
```diff
@@ -0,0 +1,123 @@
+	help
+	  Enable driver for LIS2MDL I2C-based magnetometer sensor.
+
+config LIS2MDL_NAME
```
CONFIG_LIS2MDL_I2C_MASTER_DEV_NAME, CONFIG_LIS2MDL_I2C_ADDR and LIS2MDL_I2C_MASTER_DEV_NAME will be generated from device tree "compilation".
Hence in order to prevent having them genetared twice, you should protect them under ```if !HAS_DTS_I2C_DEVICE``` directive.
You can check other drivers like LIS3MDL to check how this should look like.

erwango commented:
One remaining comment, otherwise looks fine.

mariotesi commented:
There are still some issue on this commit ?
Please let me know

erwango commented:
@galak, can you review again? I think this is ok now (binding file added)

avisconti commented:
@galak 
Can you possibly take a look at this PR?

galak reviewed (dismissed):
Can we change the GPIO for interrupt to use DT

mariotesi commented:
In the last commit device tree support has been added to GPIO interrupt selection pin/port

MaureenHelm commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -0,0 +1,342 @@
+	for (i = 0; i < LIS2MDL_NUM_AXIS; i++) {
+		offs = val->val1;
+		ret = lis2mdl_raw_write_n(lis2mdl, CONFIG_LIS2MDL_I2C_ADDR,
+					  regs, (u8_t *)&offs,
```
You assume the sensor and MCU have the same endianness. Use a byte order macro in `include/misc/byteorder.h`

MaureenHelm commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -0,0 +1,342 @@
+
+	ret = lis2mdl_raw_read_n(lis2mdl, CONFIG_LIS2MDL_I2C_ADDR,
+				 LIS2MDL_TEMP_OUT_L_REG,
+				 (u8_t *)&temp_raw, sizeof(temp_raw));
```
You should be doing the i2c transfer in a sample_fetch routine, not channel_get.

MaureenHelm commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -0,0 +1,342 @@
+	case SENSOR_CHAN_MAGN_XYZ:
+		lis2mdl_channel_get_mag(dev, chan, val);
+		return 0;
+	case SENSOR_CHAN_TEMP:
```
This should be `SENSOR_CHAN_DIE_TEMP`

MaureenHelm commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -0,0 +1,342 @@
+		return -EIO;
+	}
+
+	if (!(tmp & LIS2MDL_STS_MDA_UP)) {
```
What is this checking?

MaureenHelm commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -0,0 +1,342 @@
+{
+	struct lis2mdl_data *lis2mdl = dev->driver_data;
+
+	lis2mdl->i2c = device_get_binding(CONFIG_LIS2MDL_I2C_MASTER_DEV_NAME);
```
Store the i2c dev name in a const device config struct and get it from there.

MaureenHelm commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -0,0 +1,342 @@
+	}
+#endif
+
+	dev->driver_api = &lis2mdl_driver_api;
```
Remove this line and use `DEVICE_AND_API_INIT` below.

MaureenHelm commented on drivers/sensor/lis2mdl/lis2mdl_trigger.c:
```diff
@@ -0,0 +1,146 @@
+		lis2mdl->handler_drdy(dev, &drdy_trigger);
+	}
+
+	gpio_pin_enable_callback(lis2mdl->gpio, CONFIG_LIS2MDL_GPIO_PIN_NUM);
```
Get the pin num from the const device config struct

MaureenHelm commented on drivers/sensor/lis2mdl/lis2mdl_trigger.c:
```diff
@@ -0,0 +1,146 @@
+	struct lis2mdl_data *lis2mdl = dev->driver_data;
+
+	/* setup data ready gpio interrupt */
+	lis2mdl->gpio = device_get_binding(CONFIG_LIS2MDL_GPIO_DEV_NAME);
```
Get the gpio device name from the const device config struct

erwango commented:
@mariotesi , do you have further plans for this PR?

mariotesi commented:
I'll push CR asap

MaureenHelm reviewed (changes requested):
Please add the new sensor driver to `tests/drivers/build_all\sensors_i_z.conf` and `tests/drivers/build_all\sensors_trigger_i_z.conf` to ensure it gets built in CI.

MaureenHelm commented on drivers/sensor/lis2mdl/lis2mdl.h:
```diff
@@ -0,0 +1,148 @@
+				   u8_t mask, u8_t data);
+int lis2mdl_raw_read8(struct lis2mdl_data *lis2mdl, u8_t reg_addr, u8_t *value);
+int lis2mdl_raw_write8(struct lis2mdl_data *lis2mdl, u8_t reg_addr, u8_t value);
+int lis2mdl_raw_read_n(struct lis2mdl_data *lis2mdl, u8_t start_addr, u8_t *buf,
```
These i2c wrapper prototypes can now be removed

MaureenHelm commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -0,0 +1,376 @@
+	if (odr == 0) {
+		/* power off mag */
+		return i2c_reg_update_byte(lis2mdl->i2c,
+					   CONFIG_LIS2MDL_I2C_ADDR,
```
Get the i2c address from the device struct

MaureenHelm commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -0,0 +1,376 @@
+		return odr_idx;
+	}
+
+	return i2c_reg_update_byte(lis2mdl->i2c, CONFIG_LIS2MDL_I2C_ADDR,
```
Get the i2c address from the device struct

MaureenHelm reviewed (changes requested):
Please fix the CI failures. It looks like you need to add defines to `tests/drivers/build_all/dts.fixup`

mariotesi commented:
I don't understand why shippable test fails, my build of test/drivers/build_all everything it's ok

MaureenHelm commented:
> I don't understand why shippable test fails, my build of test/drivers/build_all everything it's ok

How are you building it locally? Did you include `-DCONF_FILE=sensors_i_z.conf` when you invoked cmake? There are two ways you should build the test:
```console
$ cmake -DBOARD=frdm_k64f -DCONF_FILE=sensors_i_z.conf ..
$ make
```
```console
$ cmake -DBOARD=frdm_k64f -DCONF_FILE=sensors_trigger_i_z.conf ..
$ make
```

MaureenHelm commented on tests/drivers/build_all/dts.fixup:
```diff
@@ -105,6 +105,16 @@
+#ifndef CONFIG_LIS2MDL_I2C_ADDR
+#define CONFIG_LIS2MDL_I2C_ADDR 0
+#define CONFIG_LIS2MDL_I2C_MASTER_DEV_NAME ""
+#endif
```
Please move the tests changes into a separate commit.

MaureenHelm commented on tests/drivers/build_all/sensors_i_z.conf:
```diff
@@ -28,3 +28,4 @@ CONFIG_TH02=y
 CONFIG_TMP007=y
 CONFIG_TMP112=y
 CONFIG_VL53L0X=y
+CONFIG_LIS2MDL=y
```
Please move this up to preserve alpha order

MaureenHelm commented on tests/drivers/build_all/sensors_trigger_i_z.conf:
```diff
@@ -25,3 +25,5 @@ CONFIG_SX9500=y
 CONFIG_TMP007=y
 CONFIG_TMP007_TRIGGER_OWN_THREAD=y
+CONFIG_LIS2MDL=y
+CONFIG_LIS2MDL_TRIGGER_OWN_THREAD=y
```
Please move up to preserve alpha order

mariotesi committed:
tests: driver: lis2mdl: Add configuration for CI tests

	Add default dts configuration for CI tests in case
	selected board used for test purpose do not export
	dts configuration for LIS2MDL

Signed-off-by: Mario Tesi <mario.tesi@st.com>
